pulp_ansible/app/tasks/collections.py

```diff
@@ -293,206 +295,219 @@ def __init__(self, remote):
+            if api_version == 2:
+                collection_versions = collection_versions_list["results"]
             else:
-                raise RuntimeError(_("Unsupported API versions at {}").format(root))
-
-            endpoint = f"{root}v{self.api_version}/collections/"
-
-            return endpoint, self.api_version
-
-        async def _get_url(page, versions_url=None):
-            if collection_info and not versions_url:
-                name, version, source = collection_info[page - 1]
-                namespace, name = name.split(".")
-                root = source or remote.url
-                api_endpoint = (await _get_collection_api(root))[0]
-                url = f"{api_endpoint}{namespace}/{name}/"
-                return url
-
-            if not versions_url:
-                api_endpoint, api_version = await _get_collection_api(remote.url)
-                return get_page_url(api_endpoint, api_version, page)
-
-            if not self.api_version:
-                await _get_collection_api(remote.url)
-
-            return get_page_url(versions_url, self.api_version, page)
-
-        async def _loop_through_pages(not_done, versions_url=None):
-            """
-            Loop through API pagination.
-            """
-            url = await _get_url(1, versions_url)
-            downloader = remote.get_downloader(url=url)
-            data = parse_metadata(await downloader.run())
-
-            count = data.get("count") or data.get("meta", {}).get("count", 1)
-            if collection_info and not versions_url:
-                count = len(collection_info)
-                page_count = count
+                collection_versions = collection_versions_list["data"]
+            for collection_version in collection_versions:
+                if collection_version["version"] in requirement_version:
+                    version_num = collection_version["version"]
+                    collection_version_detail_url = f"{collection_url}/versions/{version_num}/"
+                    if collection_metadata["deprecated"]:
+                        self.deprecations |= Q(namespace=namespace, name=name)
+                    tasks.append(
+                        asyncio.create_task(
+                            self._fetch_collection_version_metadata(
+                                api_version,
+                                collection_version_detail_url,
+                            )
+                        )
+                    )
+            next_value = self._get_response_next_value(api_version, collection_versions_list)
+            if not next_value:
+                break
+            page_num = page_num + 1
+
+        await asyncio.gather(*tasks)
+
+    @staticmethod
+    def _get_response_next_value(api_version, response):
+        if api_version == 2:
+            return response["next"]
+        else:
+            return response["links"]["next"]
```

Comment:
wondering if it can raise KeyError

Reply:
I don't expect it to for V2 or V3. If it does though, I think not catching it is right because it would be fatal.

Reply:
I agree it would be fatal, but I wonder if the error message would be clear enough

Reply:
Those keys are expected for both V2 and V3 so I don't think we need to take special care for such an exceptional situation. Generally I think exception catching is for situations where you expect to experience that exception.